src/ecs/Schedule.hpp

```diff
@@ -0,0 +1,141 @@
+/*
+** Agartha-Software, 2023
+** Cevy
+** File description:
+** Schedule
+*/
+
+#pragma once
+#include <functional>
+#include <list>
+#include <tuple>
+#include <optional>
+#include <algorithm>
+#include <typeindex>
+#include <any>
+
+#include "ecs.hpp"
+
+
+class cevy::ecs::Schedule {
+    public:
+        template<typename Before = std::nullopt_t, typename After = std::nullopt_t>
+        class Stage {
+            public:
+                template<typename T>
+                using before = Stage<T, std::nullopt_t>;
+
+                template<typename T>
+                using after = Stage<std::nullopt_t, T>;
+        };
+        template<typename T>
+        using before = Stage<>::before<T>;
+        template<typename T>
+        using after = Stage<>::after<T>;
+
+        class First : public Stage<> {};
+
+        class Update : public after<First> {};
+        class PreUpdate : public before<Update> {};
+        class PostUpdate : public after<Update> {};
+
+        class StateTransition : public before<Update> {};
+        class RunFixedUpdateLoop : public before<Update> {};
+
+        class Last : public after<PostUpdate> {};
+
+    private:
+        using type_tuple = std::tuple<std::type_index, std::any>;
+        std::list<std::type_index> _schedule;
+
+    public:
+        void init_default_schedule() {
+            insert_schedule<First>();
+            insert_schedule<Update>();
+            insert_schedule<PreUpdate>();
+            insert_schedule<PostUpdate>();
+            // insert_schedule<StateTransition>();
+            // insert_schedule<RunFixedUpdateLoop>();
+            insert_schedule<Last>();
+
+            // add systems relating to the automatic parts of
+            //StateTransitions and RunFixedUpdateLoop ?
+        }
+
+
+        template<typename T, typename T::before>
+        void insert_schedule() {
+            auto it = std::find(_schedule.begin(), _schedule.end(), std::type_index(typeid(T::before)));
+            _schedule.insert(it, std::type_index(typeid(T)));
+        }
+
+        template<typename T, typename T::after>
+        void insert_schedule() {
+            auto it = std::find(_schedule.begin(), _schedule.end(), std::type_index(typeid(T::after)));
+            ++it;
+            _schedule.insert(it, std::type_index(typeid(T)));
+        }
+
+        template<typename T>
+        void insert_schedule() {
+            _schedule.push_back(std::type_index(typeid(T)));
+        }
+
+        using system_function = std::function<void (World &)>;
+        using system = std::tuple<system_function, std::type_index>;
+        std::vector<system> _systems;
```

Comment:
these variables are hidden between all these functions. Why they are not at top of the public part ?

Reply:
J'aimerai que la pr continue malgré des détails comme ça, après je reconnais que c'est pas parfait, j'en fait une issue pour que ça ne soit pas oublié